### DISCUSSION TITLE
PIO assembly support for rp2350 enhanced version

### DISCUSSION THREAD

**Discussion by graeme-winter** [micropython/micropython#16487] [RP2040 / Pico]
With the rp2350 there are some new options to pioasm whch are not currently supported in the existing syntax e.g. `mov(rxfifo[0], isr)` - obviously you can "hack" it by hand-assembling and over-writing the right instruction e.g. with:

```
@rp2.asm_pio(
    set_init=(rp2.PIO.OUT_LOW,), sideset_init=(rp2.PIO.OUT_LOW, rp2.PIO.OUT_LOW)
)
def quad():
    # PIO assembly for quad with 32 instructions

quad[0][28] = 0x8018
```

but feels like something worth supporting if someone has the time to extend the assembler

Would probably need the state machine setup code to verify at runtime that this hardware has the right PIO version to support all the instructions given in the program e.g. if pre-compiled as a list of numbers...

**Comment by graeme-winter** [comment]
Diffs are detailed in §11.1.1. "Changes from RP2040" in the RP2350 datasheet - probably not all diffs need support in the micropython port, but some are useful

**Comment by mattytrentini** [comment]
For reference, most of the PIO assembly compilation occurs in Python, in [ports/rp2/modules/rp2.py](https://github.com/micropython/micropython/blob/master/ports/rp2/modules/rp2.py). 

It's relatively understandable code in case anyone wants to consider extending it with the newer features!

**Comment by magy00** [comment]
See also:
[https://forums.raspberrypi.com/viewtopic.php?t=375677](https://forums.raspberrypi.com/viewtopic.php?t=375677)
[https://github.com/raspberrypi/pico-sdk/issues/2166](https://github.com/raspberrypi/pico-sdk/issues/2166)
